lib/utils/replace.go

```diff
@@ -221,6 +221,110 @@ func KubeResourceMatchesRegex(input types.KubernetesResource, resources []types.
+	if len(input.Verbs) != 1 {
+		return false, trace.BadParameter("only one verb is supported, input: %v", input.Verbs)
+	}
+	if input.Name != "" {
+		return false, trace.BadParameter("name is not supported for KubeResourceMatchesNamespaceRegex")
+	}
+
+	verb := input.Verbs[0]
+	isDeny := cond == types.Deny
+
+	// isClusterWideResource is true if the resource is cluster-wide, e.g. a
+	// namespace resource or a clusterrole.
+	isClusterWideResource := slices.Contains(types.KubernetesClusterWideResourceKinds, input.Kind)
+
+	// If the user is list/read/watch a namespace, they should be able to see the
+	// namespace they have resources defined for.
+	// This is a special case because we don't want to require the user to have
+	// access to the namespace resource itself.
+	// This is only allowed for the list/read/watch verbs because we don't want
+	// to allow the user to create/update/delete a namespace they don't have
+	// permissions for.
+	targetsReadOnlyNamespace := input.Kind == types.KindKubeNamespace &&
+		slices.Contains([]string{types.KubeVerbGet, types.KubeVerbList, types.KubeVerbWatch}, verb)
+
+	for _, resource := range resources {
+		// If the resource has a wildcard verb, it matches all verbs.
+		// Otherwise, the resource must have the verb we're looking for otherwise
+		// it doesn't match.
+		// When the resource has a wildcard verb, we only allow one verb in the
+		// resource input.
+		if !isVerbAllowed(resource.Verbs, verb) {
+			continue
+		}
+		switch {
+		// If the user has access to a specific namespace, they should be able to
+		// access all resources in that namespace.
+		case resource.Kind == types.KindKubeNamespace:
+			cond := !isDeny || resource.Name == types.Wildcard
+			if input.Namespace == "" && cond {
+				return cond, nil
+			}
+			// Access to custom resources is determined by the access level of the
+			// namespace resource where the custom resource is defined.
+			// This is a special case because custom resources are not defined in the
+			// user's resources list.
+			// Access to namspaced resources is determined by the access level of the
+			// namespace resource where the resource is defined or by the access level
+			// of the resource if supported.
+			if ok, err := MatchString(input.Namespace, resource.Name); err != nil || ok && cond {
+				return cond || isDeny, trace.Wrap(err)
+			}
+		case targetsReadOnlyNamespace && resource.Kind != types.KindKubeNamespace && resource.Namespace != "":
+			// If the user requests a read-only namespace get/list/watch, they should
+			// be able to see the list of namespaces they have resources defined in.
+			// This means that if the user has access to pods in the "foo" namespace,
+			// they should be able to see the "foo" namespace in the list of namespaces
+			// but only if the request is read-only.
+			cond := !isDeny || resource.Name == types.Wildcard && resource.Namespace == types.Wildcard
+			if input.Namespace == "" && cond {
+				return cond, nil
+			}
```

Comment:
Here, input kind is a namespace (because of `targetsReadOnlyNamespace`). Is there a case in which `input.Namespace` is not `""`? I thought kube did not support nested namespaces.

Reply:
You are correct. I copied from other location and forgot to think :disappointed: